pep-3157.txt

```diff
@@ -0,0 +1,247 @@
+        Batuhan Taskaya <batuhan@python.org>,
+        Lysandros Nikolaou <lisandrosnik@gmail.com>
+Sponsor: Pablo Galindo <pablogsal@python.org>
+Discussions-To: python-dev@python.org
+Type: Standards Track
+Status: Draft
+Content-Type: text/x-rst
+Python-Version: 3.9
+Created: 15-Nov-2021
+
+
+Abstract
+========
+
+This document proposes to lift some of the restrictions originally formulated in
+:pep:`498` and to provide a formalized grammar for f-strings that can be
+integrated into parser directly. The proposed syntactic formalization of
+f-strings will have some small side-effects over how f-strings are parsed and
+interpreted, allowing for a considerable number of advantages for end users and
+library developers, while also improving dramatically the maintainance cost of
+the code dedicated to parse f-strings.
+
+
+Motivation
+==========
+
+When f-strings were originally introduced in :pep:`398` the specification was
+provided without providing a formal Grammar for f-strings. Additionally, theGG
+specification contains several restrictions that are imposed so the parsing of
+f-strings could be implemented into CPython without modifying the existing
+lexer. These limitations have been recognized previously and previous attempt
+have been made to lift them in :pep:`536` but non of this work was ever implemented 
+(see [2]_). Some of these limitations (collected originally by :pep:`536`) are:
+
+#. It is impossible to use the quote character delimiting the f-string
+   within the expression portion::
+
+    >>> f'Magic wand: { bag['wand'] }'
+                                 ^
+    SyntaxError: invalid syntax
+
+#. A previously considered way around it would lead to escape sequences
+   in executed code and is prohibited in f-strings::
+
+    >>> f'Magic wand { bag[\'wand\'] } string'
+    SyntaxError: f-string expression portion cannot include a backslash
+
+#. (TODO: DO WE WANT THIS?) Comments are forbidden even in multi-line f-strings::
+
+    >>> f'''A complex trick: {
+    ... bag['bag']  # recursive bags!
+    ... }'''
+    SyntaxError: f-string expression part cannot include '#'
+
+#. (TODO: DO WE WANT THIS?) Expression portions need to wrap ``':'`` and ``'!'`` in braces::
+
+    >>> f'Useless use of lambdas: { lambda x: x*2 }'
+    SyntaxError: unexpected EOF while parsing
+
+#. Arbitrary nesting of expressions without expansion of escape sequences is
+   available in every single other language employing a string interpolation
+   method that uses expressions instead of just variable names. [6]_
+
+These limitations serve no purpose from a language user perspective and
+can be lifted by giving f-literals a regular grammar without exceptions
+and implementing it using dedicated parse code.
+
+The other problems that f-strings have is that the current implementation in
+CPython relies on the existing string parsing machinery and a post processing of
+its tokens. This has the following problems:
```

Comment:
I guess the point here is that f-string get tokenised as `STRING` tokens, but I had to go through the sentence twice to understand it. Maybe something like this would be better:

```suggestion
The other problems that f-strings have is that the current implementation in
CPython relies on tokenising f-string as `STRING` tokens and a post processing of
these tokens. This has the following problems:
```